test/test_common/environment.h

```diff
@@ -72,6 +72,47 @@ class TestEnvironment {
    */
   static std::string getCheckedEnvVar(const std::string& var);
 
+  /**
+   * Generates an appropriate base-id for use as the base id option to an Envoy
+   * server. Each test that requires a unique test base id should invoke this
+   * method with a unique value to get a string value appropriate for use the
+   * value of the --base-id command line argument. In general, tests that
+   * create an Envoy::Server::HotRestartImpl without mocks should use this
+   * function. If all test cases within a grouping, say a single test source
+   * file, are executed consecutively, they may share a test base id. Tests in
+   * separate groupings cannot share a test base id for reasons described
+   * below. Special care must be taken with death tests -- the gtest framework
+   * will fork and invoke a new copy of the same test binary with additional
+   * command line flags. This can result in the re-use of a given base-id from
+   * another process (which causing a failure).
+   *
+   * We require a unique test base id because random seeds are reused across
+   * multiple test targets. For example, running:
+   *     bazel --runs_per_test=3 //test:foo_test //test:bar_test
+   * results in 3 unique seeds. Each test target is run once with each seed.
+   * Similarly, in coverage tests, the same test may be run concurrently in
+   * multiple coverage shards.
+   *
+   * This method uses the given test base id and one of the following
+   * environment variables:
+   * - TEST_RANDOM_SEED is used to handle concurrent runs via the Bazel
+   *                    --runs_per_test flag
+   * - TEST_SHARD_INDEX is used to handle concurrent runs when tests are run in
+   *                    shards, such as in coverage testing
+   *
+   * This algorithm is re-implemented in the following test scripts:
+   * - test/integration/hot_restart_test.sh
+   * - test/integration/run_envoy_test.sh
```

Comment:
I shudder to even suggest this, but given how confusing/complicated this is already, should we just implement this "algorithm" as a `.sh` script and then just shell out to it when we need to from the CC file? Then all the logic is in one place with good comments. Up to you.

Reply:
It still requires everyone to pick a unique set of test base ids. 

Sigh. I think the gold-plated solution is probably some special value for base-id that tells Envoy to pick an unused base id (like creating a socket on port 0) and a flag to write the resulting base id to a file. Most tests don't need to know the base id -- they just need to not collide with other envoy processes and the test that does need to know (hotrestart_test.sh) can get the base id and use it for its testing.

Reply:
Yeah this is all terrible. I'm fine with this for now given the comments as it is strictly better than what we have now, but up to you.

Reply:
Oh, you know me. Can't resist a challenge. If you want to merge this I'll put up a PR with the better fix.

Reply:
Your call. :)